Source/NujasCombat/Private/DynamicTargetingComponent.cpp

```diff
@@ -316,32 +316,59 @@ TArray<AActor*> UDynamicTargetingComponent::FindAllActorsOnScreen()
 
 void UDynamicTargetingComponent::FindTargetWithAxisInput(float AxisInput)
 {
-	if(!ArrowComponent || !SelectedActor || FMath::Abs(AxisInput) > START_AXIS_THRESHOLD)
+	if(!ArrowComponent || !SelectedActor || FMath::Abs(AxisInput) < START_AXIS_THRESHOLD)
+	{
+		AxisCache = 0.f;
+		return;
+	}
+	if(AxisCache * AxisInput > 0.f)
 		return;
+	AxisCache = AxisInput;
 	TArray<AActor*> AllActors = FindAllActorsOnScreen();
+
 	bool bSuccessfulProjection = false;
-	const FVector2D SelectedActorScreenPosition = UViewportUtility::GetActorOnScreenPosition(SelectedActor, PlayerController, bSuccessfulProjection);
-	float ClosestDistanceToSelectedActor = BIG_NUMBER;
-	AActor* ClosestActor = nullptr;
-	for (AActor*& Actor : AllActors)
+	AActor* FoundNextTarget = nullptr;
+	const FVector2D OwnerScreenProjection = UViewportUtility::GetActorOnScreenPosition(Owner, PlayerController, bSuccessfulProjection);
+	float ClosestEnemyDistance = NAN;
+	float ClosestEnemyX = NAN;
+	if(!bSuccessfulProjection) return;
+
+	for (AActor*& PotentialTarget : AllActors)
 	{
-		if(Actor == SelectedActor)
-			continue;
 
-		const FVector2D ScreenPosition = UViewportUtility::GetActorOnScreenPosition(Actor, PlayerController, bSuccessfulProjection);
-		const float DistanceFromSelectedActor = SelectedActorScreenPosition.X - ScreenPosition.X;
-		if(FMath::Sign(AxisInput) == FMath::Sign(DistanceFromSelectedActor))
+		if(PotentialTarget == SelectedActor)
+			continue;
+		const float DistSquared = FVector::DistSquared(PotentialTarget->GetActorLocation(), Owner->GetActorLocation());
+		if(DistSquared > MaxDistanceToTargetSquared || DistSquared < MinDistanceToTargetSquared)
+			continue;
+		// no need to check if projection worked, because the target has already entered into the onscreen targets list
+		const FVector2D PotentialTargetScreenPosition = UViewportUtility::GetActorOnScreenPosition(SelectedActor, PlayerController, bSuccessfulProjection);
+		const float XOffset = PotentialTargetScreenPosition.X - OwnerScreenProjection.X;
+		if(AxisInput * XOffset > 0.f)
 		{
-			const float AbsDistanceFromSelectedActor = FMath::Abs(DistanceFromSelectedActor);
-			if(AbsDistanceFromSelectedActor < ClosestDistanceToSelectedActor)
+			if(FoundNextTarget)
 			{
-				ClosestDistanceToSelectedActor = AbsDistanceFromSelectedActor;
-				ClosestActor = Actor;
+				if(DistSquared < ClosestEnemyDistance || FMath::Abs(XOffset) < FMath::Abs(ClosestEnemyX))
+				{
+					ClosestEnemyDistance = DistSquared;
```

Comment:
353-355 abd 360-362 are essentially the same code path. Should trim this condition. Seems equivalent to this (un-tested tho):

```
if(AxisInput * XOffset > 0.f && 
       !FoundNextTarget || 
       DistSquared < ClosestEnemyDistance || 
       FMath::Abs(XOffset) < FMath::Abs(ClosestEnemyX
) {
		ClosestEnemyDistance = DistSquared;
		ClosestEnemyX = XOffset;
		FoundNextTarget = PotentialTarget;
}

```